### PR TITLE
[BUGFIX] Ensure owner is available during resolution for all components

### PR DESCRIPTION
**Pull request by pzuraq**
Previously, strict components were not passed an owner when they were
defined. This turned out to be incorrect, because even if strict
components did not use an owner to resolve components, the components
that they used in scope could potentially be non-strict, and resolve
components. Those components effectively become different components
based on the currently active owner, so we have to pass the owner no
matter what type of component we are rendering.

### CODE REVIEW

File: packages/@glimmer/integration-tests/test/owner-test.ts [hunk 1]
```diff
@@ -81,6 +169,34 @@ class OwnerTest extends RenderTest {
 
     assert.verifySteps(['foo-bar owner called', 'foo-baz owner called'], 'owners used correctly');
   }
+
+  @test
+  'owner can be changed by a component with the hasSubOwner capability'(assert: Assert) {
```
Comment (pzuraq) [inline packages/@glimmer/integration-tests/test/owner-test.ts:174]: This test is unrelated, I just realized we didn't have it (or at least not an obvious one for it) so wanted to add it

**Review by krisselden — CHANGES_REQUESTED** [review]
Can you add a test that passes a closure component into the model property of the mount and invoke that component inside of the mount with a yield block that invokes a component coming from the mount?

```
{{mount 'engine' model=MyComponent}}
```

inside the mounted 

```
{{#let @model as |OuterComponent|}}
   <OuterComponent>
      <InnerComponent/>
  </OuterComponent>
{{/let}}
```

I want to ensure you didn't conflate the owner (realm) concept with dynamic scope.

File: packages/@glimmer/program/lib/constants.ts [hunk 2]
```diff
@@ -196,15 +196,16 @@ export class ConstantsImpl
...
   component(
     definitionState: ComponentDefinitionState,
+    object: object,
     isOptional: true
   ): ComponentDefinition | null;
   component(
     definitionState: ComponentDefinitionState,
-    isOptional?: true,
-    owner?: object
+    owner: object,
```
Comment (krisselden) [inline packages/@glimmer/program/lib/constants.ts:207]: if owner is optional shouldn't this allow undefined?

Comment (krisselden) [reply packages/@glimmer/program/lib/constants.ts:207]: I don't really understand this overload

**Review by krisselden — APPROVED** [review]

File: packages/@glimmer/integration-tests/test/owner-test.ts [hunk 3]
```diff
@@ -81,6 +169,73 @@ class OwnerTest extends RenderTest {
...
+
+    const Mount2 = defineMountComponent(owner2, { CheckOwner2 }, `<CheckOwner2/><@CheckOwner2/>`);
+    const Mount1 = defineMountComponent(
+      owner1,
+      { CheckOwner1, CheckOwner2, Mount2 },
+      `<CheckOwner1/><Mount2 @CheckOwner2={{CheckOwner2}}/>`
+    );
+
+    this.renderComponent(Mount1);
+  }
```
Comment (rwjblue) [inline packages/@glimmer/integration-tests/test/owner-test.ts:228]: For context, we (@pzuraq + @krisselden + myself) discussed this in detail. The rough conclusion is that we don't really like the thing this test is asserting (it is very strange to end up with an owner from a different invocation than where you were originally associated), but it is quite difficult to ensure that we **always** have the right owner (it isn't even possible to absorb all variants of things that could be done in JavaScript land).

At the moment, we intend to land this update over in Ember and add a debug only AST transform that wraps any values passed to `{{mount 'foo' model=` (because as far as we can tell, that is the only way to pass things "across owner / realm") and asserts if you pass something that is a component/helper/modifier that isn't wrapped in `(component`/`(helper`/`(modifier` helpers.

File: packages/@glimmer/program/lib/constants.ts [hunk 4]
```diff
@@ -196,15 +196,11 @@ export class ConstantsImpl
...
   component(
     definitionState: ComponentDefinitionState,
-    isOptional: true
-  ): ComponentDefinition | null;
-  component(
-    definitionState: ComponentDefinitionState,
-    isOptional?: true,
-    owner?: object
+    owner: object,
+    isOptional?: true
```
Comment (krisselden) [inline packages/@glimmer/program/lib/constants.ts:203]: isOptional?: true should be boolean, I don't think the narrowness helps here.
